### DISCUSSION TITLE
Adding support for deepinfra APIs (or perhaps a plugin support to integrate facilities of choice)

### DISCUSSION THREAD

**Discussion by unlucio** [open-webui/open-webui#5005] [Ideas]
As someone who doesn't have the hardware to run local models in their infrastructure, I've found `deepinfra` to be a convenient solution. Their OpenAI-compliant API has allowed me to seamlessly integrate `open-webui` with their LLMs for my text generation needs.

What's more, `deepinfra` offers a range of capabilities beyond text generation, including Text To Image, Automatic Speech Recognition, and embeddings. Eager to leverage these features, I set out to integrate voice and image capabilities into my workflow. However, I discovered that `deepinfra` doesn't expose an OpenAI-compliant API for models beyond text generation.

Despite this, their documentation is excellent, and they provide Node.js modules that make integration feasible. I'm excited about the prospect of adding support for these additional capabilities and contributing to the community.

As suggested in the `open-webui` contributing guidelines (https://docs.openwebui.com/contributing), I'd like to gauge interest in this project before proceeding. If there's sufficient interest, I'd be happy to take the lead on implementing OpenAI-compliant API support for `deepinfra`'s non-text generation models.

Let me know if this is something you'd like to see happen!

**Comment by unlucio** [comment]
I noticed [this post](https://github.com/open-webui/open-webui/discussions/4919) about another PAI integration.
Perahps the better answer would be to have a plugin system and the final user chooses their preferred backend facility?

**Comment by unlucio** [comment]
Meanwhile, I did a little POC:

![screencapture-localhost-8080-c-9ba941ea-27ee-4b70-83c2-18c2be62686a-2024-08-30-18_01_26](https://github.com/user-attachments/assets/9c6f0601-d58b-46e0-b798-ea2b4a372a54)
<img width="1029" alt="Screenshot 2024-08-30 at 17 49 04" src="https://github.com/user-attachments/assets/f209ef66-226d-47bd-a9f1-32462ee8eacf">
<img width="512" alt="Screenshot 2024-08-30 at 18 02 53" src="https://github.com/user-attachments/assets/70d6b0db-2767-4a5f-84ac-0357056be152">

**Comment by justinh-rahb** [comment]
You're looking for Functions:

https://docs.openwebui.com/tutorial/functions
https://openwebui.com/functions

**Comment by JustAlternate** [comment]
Hi, are you still using deepinfra ?

I have an issue where R1 stop after a certain amount a token

and I tested the openai API like of deepinfra and the max_tokens parameter look like is working fine when doing request by hand.

But changing the "max_tokens" parameter in open web ui in "advanced parameters" for my prompt is not working at all 

Setting it to 1 for example is not giving me 1 token response, and this problem is for all models

**Reply by JustAlternate** [reply]
@unlucio

**Reply by unlucio** [reply]
I'm using deep infra and R1 but I'm not having any issues with it

**Reply by JustAlternate** [reply]
yes it was fixed in 5.8 sorry for the ping :)

**Comment by LucianU** [comment]
For anyone not familiar enough with the ecosystem, you add the Deepinfra connection at Admin Settings > Connections under the OpenAI API section. Currently, their endpoint is `https://api.deepinfra.com/v1/openai`.
